### PR TITLE
Fix NAR tests on Linux+ZFS+normalize

### LINKED ISSUE #11595: The nar.sh checks fail for some Linux machines
## Platform

- [x] Linux: 
- [ ] macOS
- [ ] WSL

## Additional information

When building nix as a dependency of another project through the `github:nixos/nix` flake on a Linux runner, latest Nix (I've tried several commits, last was 322d2c767f2a3f8ef2ac3d1ba46c19caf9a1ffce) fails to build (more specifically, the tests fail). The host is a Zen 4 EPYC with x86_64 Linux (NixOS 24.05). Filesystem is ZFS.

On another local Linux machine (`Linux 6.10.11-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.10.11-1 (2024-09-22) x86_64 GNU/Linux`), the project builds correctly.

The failing test is `nar.sh`, more specifically this part:

https://github.com/NixOS/nix/blob/master/tests/functional/nars.sh#L91-L100

It seems that Darwin and Linux have a different expected behaviour here, but for some reason the Linux runner specifically ends up having Darwin's behaviour.

## Output

<details><summary>Output</summary>

```log

nix>     +(nars.sh:94) [[ Linux = Darwin ]]
nix>     +(nars.sh:97) nix-store --restore /build/nix-test/nars/out
nix>     error: path '/build/nix-test/nars/out/â' already exists
nix>     ++(nars.sh:97) onError
nix>     ++(/build/source/tests/functional/common/functions.sh:237) set +x
nix>     nars.sh: test failed at:
nix>       main in nars.sh:97

```

</details>

Full output available here: https://github.com/tweag/nickel/actions/runs/11016505458/job/30592178626?pr=2037#step:5:3894

## Priorities

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by yannham**
Closes #11595.

## Motivation & context

A test added recently checks that when trying to deserialize a NAR with two files that Unicode-normalize to the same result either succeeds on Linux, or fails with an "already exists" error on Darwin. However, failing with an "already exists" error can in fact also happen on Linux, when using ZFS with the proper utf8 and Unicode normalization options set.

This commit fixes the issue by not assuming the behavior from the current system, but just by blindly checking that either one of the two aforementioned possibilities happen, whether on Darwin or on Linux.

This has been tested on a Linux+ZFS+normalize runner here: https://github.com/tweag/nickel/actions/runs/11073351502/job/30769680803?pr=2037

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by Mic92 — APPROVED** [review]

**Review by Gerg-L — APPROVED** [review]
works on my machine with normalization set to `formD` lgtm

**Comment by puckipedia** [comment]
There's a bit of an issue with this actually; UTF-8 normalisation means that nix builds done on this system may inherently end up with a different NAR hash than those done on a system without normalisation. This may also present as nar files from `cache.nixos` not having the same nar hash.

**Comment by tomberek** [comment]
> There's a bit of an issue with this actually; UTF-8 normalisation means that nix builds done on this system may inherently end up with a different NAR hash than those done on a system without normalisation. This may also present as nar files from `cache.nixos` not having the same nar hash.

Is that an issue with this test? or with NAR deserialization?

**Comment by Mic92** [comment]
Search for the `use-case-hack` option, that is used as a mitigation.

**Comment by roberth** [comment]
To summarize the underlying issue:

- You may need a similar hack to the case hack in order to use such a store transparently as a vessel for deployments to stores that do not have this normalization.
- This is also an insidious [build impurity](https://hydra.nixos.org/build/274162798/download/1/manual/glossary.html#gloss-purity)  (and those are a fact of life)
  
For now, let's make the tests pass.

We could gather related ideas in
- #11555

File: tests/functional/nars.sh [hunk 1]
```diff
@@ -88,15 +88,30 @@ touch "$TEST_ROOT/case/xt_CONNMARK.h~nix~case~hack~3"
...
-if [[ $(uname) = Darwin ]]; then
-    expectStderr 1 nix-store --restore "$TEST_ROOT/out" < unnormalized.nar | grepQuiet "path '.*/out/â' already exists"
-else
-    nix-store --restore "$TEST_ROOT/out" < unnormalized.nar
+# Deserializing a NAR that contains file names that Unicode-normalize to the
+# same name should fail on macOS and specific Linux setup (typically ZFS with
+# `utf8only` enabled and `normalization` set to anything else than `none`).
+# The deserialization should succeed on most Linux.
+#
+# We test that it either succeeds or fails with "already exists" error.
```
Comment (roberth) [inline tests/functional/nars.sh:96]: Could we determine the properties of the `$TEST_ROOT`'s file system by creating these semi-duplicate files using other means? That way can at least still test that the Nix behavior matches another program's behavior, and for example catch a bad change that causes Nix to normalize on its own accord on a file system where it shouldn't.

Something like
```bash
touch $TEST_ROOT/poll-unicode-â
touch $TEST_ROOT/poll-unicode-â
n=$(wc -l $(ls $TEST_ROOT | grep poll-unicode-))
normalizing=$((n == 1))
```

**Review by roberth — CHANGES_REQUESTED** [review]

File: tests/functional/nars.sh (hunk 1 above)
Comment (yannham) [reply tests/functional/nars.sh:96]: I imagine the second `touch` would error out, in the same way as `nix-store --restore`. But we can still test that `touch` and `nix-store --restore` have the same result, at least :+1:

**Comment by yannham** [comment]
I implemented @roberth 's suggestion in the latest force push, tested again on a Linux+ZFS+utf8 normalization runner. I also changed the tests of the return code in the `if`-`then`-`else` from `[[ $code == n ]]` to `(( code == n ))`, which I believe is more appropriate.

**Comment by yannham** [comment]
Dear maintainers, is there anything else to be done here? It's probably not a very important change but I fear that if it doesn't move forward it's going to be forgotten for ever :sweat_smile: pinging @roberth once again (sorry if inappropriate)

**Review by Mic92 — APPROVED** [review]

**Review by Gerg-L — APPROVED** [review]
Works on my system with formD

**Review by bryanhonof — APPROVED** [review]

**Review by roberth — APPROVED** [review]
I might have written the conditional on `touchFilesCount` instead, but it'd be equivalent. :+1:

**Pull request merged**